File: python/rspm/render.py
"""Render rspm task tables and task logs."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta, timezone, tzinfo
from zoneinfo import ZoneInfo

from rspm.client import TaskInfo

TASK_NAME_WIDTH = 32


@dataclass(frozen=True)
class RenderLogOptions:
    """Filtering options used by log rendering helpers."""

    lines: int | None = None
    grep: str | None = None
    since: datetime | None = None


def format_task_table(tasks: list[TaskInfo]) -> str:
    """Render tasks using the same table style as the rspm CLI."""

    rows = [
        (
            f"{'TASK_ID':<8} {'NAME':<32} {'MODE':<10} {'PID':<8} {'STATUS':<12} "
            f"{'HEALTH':<8} {'RESTARTS':<8} {'UPTIME':<10} {'START_TIME':<15} "
            f"{'STOP_TIME':<15} {'CPU':<8} {'MEM':<8} NEXT"
        )
    ]
    for task in tasks:
        pid = str(task.pid) if task.pid is not None else "-"
        uptime = format_duration(task.uptime_ms) if task.uptime_ms is not None else "-"
        started = _format_task_time(task.started_at, task.display_timezone)
        stopped = _format_task_time(task.stopped_at, task.display_timezone)
        cpu_text = f"{task.cpu_percent:.1f}%" if task.cpu_percent is not None else "-"
        rows.append(
            f"{task.task_id:<8} {_fixed_width_cell(task.name, TASK_NAME_WIDTH)} "
            f"{_display_run_mode(task):<10} {pid:<8} "
            f"{_colored_status_cell(task.status)} {_colored_health_cell(task.health or '-')} "
            f"{_colored_restarts_cell(task.restart_count)} {uptime:<10} {started:<15} "
            f"{stopped:<15} {_colored_cpu_cell(cpu_text)} {_colored_memory_cell(task.memory_bytes)} "
            f"{task.schedule_state or '-'}"
        )
    rows.append(_colored_note_line(f"Timezone: {_table_display_timezone(tasks)}"))
    return "\n".join(rows) + "\n"


def format_prefixed_logs(
    task: str,
    logs: str,
    options: RenderLogOptions | None = None,
) -> str:
    """Render log lines with ``task | `` prefixes while preserving ANSI styles."""

    options = options or RenderLogOptions()
    output = "".join(f"{task} | {line}" for line in _selected_log_lines(logs, options))
    if logs and not logs.endswith("\n"):
        output += "\n"
    return output


def format_merged_logs(
    logs: list[tuple[str, str]],
    options: RenderLogOptions | None = None,
) -> str:
    """Render aggregate logs ordered by parseable RFC3339 timestamps."""

    options = options or RenderLogOptions()
    lines: list[tuple[datetime | None, int, str, str]] = []
    sequence = 0
    for task, text in logs:
        for line in _selected_log_lines(text, options):
            lines.append((_line_timestamp(line), sequence, task, line))
            sequence += 1
    lines.sort(key=lambda item: (item[0] is None, item[0] or datetime.max, item[1]))
    return "".join(f"{task} | {line}" for _, _, task, line in lines)


def format_duration(uptime_ms: int) -> str:
    """Format milliseconds as compact seconds, minutes, hours, or days."""

    seconds = uptime_ms // 1_000
    if seconds < 60:
        return f"{seconds}s"
    minutes = seconds // 60
    if minutes < 60:
        return f"{minutes}m{seconds % 60}s"
    hours = minutes // 60
    if hours < 24:
        return f"{hours}h{minutes % 60}m"
    days = hours // 24
    return f"{days}d{hours % 24}h"


def format_bytes(value: int) -> str:
    """Format bytes as B, KB, MB, or GB."""

    kb = 1024
    mb = 1024 * kb
    gb = 1024 * mb
    if value >= gb:
        return f"{value // gb}GB"
    if value >= mb:
        return f"{value // mb}MB"
    if value >= kb:
        return f"{value // kb}KB"
    return f"{value}B"


def _fixed_width_cell(value: str, width: int) -> str:
    if len(value) > width:
        value = value[: width - 3] + "..."
    return f"{value:<{width}}"


def _selected_log_lines(logs: str, options: RenderLogOptions) -> list[str]:
    lines = [
        line
        for line in logs.splitlines(keepends=True)
        if (options.grep is None or options.grep in line)
        and (options.since is None or (_line_timestamp(line) or datetime.min) >= options.since)
    ]
    if options.lines is not None:
        lines = lines[-options.lines :]
    return lines


def _line_timestamp(line: str) -> datetime | None:
    for token in line.split():
        stripped = token.strip("[](),")
        if stripped.endswith("Z"):
            stripped = stripped[:-1] + "+00:00"
        try:
            return datetime.fromisoformat(stripped).astimezone(timezone.utc)
        except ValueError:
            continue
    return None


def _format_task_time(value: str | None, display_timezone: str | None) -> str:
    if value is None:
        return "-"
    parsed = value[:-1] + "+00:00" if value.endswith("Z") else value
    dt = datetime.fromisoformat(parsed).astimezone(_display_tz(display_timezone))
    return dt.strftime("%m-%d %H:%M:%S")


def _display_tz(display_timezone: str | None) -> tzinfo:
    if display_timezone is None or display_timezone == "local":
        return datetime.now().astimezone().tzinfo or timezone.utc
    if display_timezone.startswith("UTC") or display_timezone.startswith("GMT"):
        suffix = display_timezone[3:]
        if not suffix:
            return timezone.utc
        sign = 1 if suffix[0] == "+" else -1
        hours, _, minutes = suffix[1:].partition(":")
        return timezone(sign * timedelta(hours=int(hours), minutes=int(minutes or 0)))
    return ZoneInfo(display_timezone)


def _display_run_mode(task: TaskInfo) -> str:
    return task.run_mode or "-"


def _table_display_timezone(tasks: list[TaskInfo]) -> str:
    for task in tasks:
        if task.display_timezone:
            return task.display_timezone
    return "local"


def _colored_status_cell(status: str) -> str:
    return _colorize_status(status, f"{status:<12}")


def _colorize_status(status: str, value: str) -> str:
    if status in {"online", "healthy"}:
        color = "32"
    elif status in {"unhealthy", "failed", "stopped"}:
        color = "31"
    elif status in {"starting", "scheduled", "waiting", "stopping"}:
        color = "33"
    elif status == "backoff":
        color = "35"
    else:
        color = "90"
    return f"\x1b[{color}m{value}\x1b[0m"


def _colored_health_cell(health: str) -> str:
    value = f"{health:<8}"
    if health == "ok":
        return f"\x1b[32m{value}\x1b[0m"
    if health == "fail":
        return f"\x1b[31m{value}\x1b[0m"
    if health == "-":
        return value
    return f"\x1b[33m{value}\x1b[0m"


def _colored_restarts_cell(restart_count: int) -> str:
    value = f"{restart_count:<8}"
    return f"\x1b[33m{value}\x1b[0m" if restart_count >= 3 else value


def _colored_cpu_cell(cpu: str) -> str:
    value = f"{cpu:<8}"
    try:
        high = float(cpu.rstrip("%")) >= 80.0
    except ValueError:
        high = False
    return f"\x1b[33m{value}\x1b[0m" if high else value


def _colored_memory_cell(memory_bytes: int | None) -> str:
    text = format_bytes(memory_bytes) if memory_bytes is not None else "-"
    value = f"{text:<8}"
    if memory_bytes is not None and memory_bytes >= 512 * 1024 * 1024:
        return f"\x1b[33m{value}\x1b[0m"
    return value


def _colored_note_line(value: str) -> str:
    return f"\x1b[90m{value}\x1b[0m"

File: crates/rspm-sdk/src/render.rs
use std::fmt::Write as _;

use chrono::{DateTime, Utc};
use rspm_core::config::{ProjectConfig, RestartPolicy, TaskConfig};
use rspm_core::display::format_display_table_time;
use rspm_core::state::{TaskInfo, TaskStatus};

const TASK_NAME_WIDTH: usize = 32;

#[derive(Debug, Clone, Default)]
pub struct RenderLogOptions {
    pub lines: Option<usize>,
    pub grep: Option<String>,
    pub since: Option<DateTime<Utc>>,
}

#[derive(Debug)]
struct LogLine {
    task: String,
    line: String,
    timestamp: Option<DateTime<Utc>>,
    sequence: usize,
}

pub fn format_task_table(tasks: &[TaskInfo]) -> String {
    let mut output = String::new();
    write_task_table(&mut output, tasks).expect("writing to String cannot fail");
    output
}

pub fn write_task_table(output: &mut impl std::fmt::Write, tasks: &[TaskInfo]) -> std::fmt::Result {
    writeln!(
        output,
        "{:<8} {:<32} {:<10} {:<8} {:<12} {:<8} {:<8} {:<10} {:<15} {:<15} {:<8} {:<8} NEXT",
        "TASK_ID",
        "NAME",
        "MODE",
        "PID",
        "STATUS",
        "HEALTH",
        "RESTARTS",
        "UPTIME",
        "START_TIME",
        "STOP_TIME",
        "CPU",
        "MEM"
    )?;
    for task in tasks {
        let status = colored_status_cell(task.status);
        let health = colored_health_cell(task.health.as_deref().unwrap_or("-"));
        let restarts = colored_restarts_cell(task.restart_count);
        let cpu_text = task.cpu_percent.map(format_cpu_percent);
        let cpu = colored_cpu_cell(cpu_text.as_deref().unwrap_or("-"));
        let memory = colored_memory_cell(task.memory_bytes);
        let pid = task
            .pid
            .map(|pid| pid.to_string())
            .unwrap_or_else(|| "-".to_string());
        let uptime = task
            .uptime_ms
            .map(format_duration)
            .unwrap_or_else(|| "-".to_string());
        let started = task
            .started_at
            .as_ref()
            .map(|time| format_display_table_time(time, task.display_timezone.as_deref()))
            .unwrap_or_else(|| "-".to_string());
        let stopped = task
            .stopped_at
            .as_ref()
            .map(|time| format_display_table_time(time, task.display_timezone.as_deref()))
            .unwrap_or_else(|| "-".to_string());
        writeln!(
            output,
            "{:<8} {:<32} {:<10} {:<8} {} {} {} {:<10} {:<15} {:<15} {} {} {}",
            task.task_id,
            fixed_width_cell(&task.name, TASK_NAME_WIDTH),
            display_run_mode(task),
            pid,
            status,
            health,
            restarts,
            uptime,
            started,
            stopped,
            cpu,
            memory,
            task.schedule_state.as_deref().unwrap_or("-")
        )?;
    }
    writeln!(
        output,
        "{}",
        colored_note_line(&format!("Timezone: {}", table_display_timezone(tasks)))
    )
}

pub fn format_offline_task_table(config: &ProjectConfig) -> String {
    let mut output = String::new();
    write_offline_task_table(&mut output, config).expect("writing to String cannot fail");
    output
}

pub fn write_offline_task_table(
    output: &mut impl std::fmt::Write,
    config: &ProjectConfig,
) -> std::fmt::Result {
    writeln!(
        output,
        "{:<8} {:<32} {:<10} {:<8} {:<12} {:<8} {:<8} {:<10} {:<15} {:<15} {:<8} {:<8} NEXT",
        "TASK_ID",
        "NAME",
        "MODE",
        "PID",
        "STATUS",
        "HEALTH",
        "RESTARTS",
        "UPTIME",
        "START_TIME",
        "STOP_TIME",
        "CPU",
        "MEM"
    )?;
    for (index, (task_name, task)) in config.tasks.iter().enumerate() {
        writeln!(
            output,
            "{:<8} {:<32} {:<10} {:<8} {:<12} {:<8} {:<8} {:<10} {:<15} {:<15} {:<8} {:<8} -",
            index + 1,
            fixed_width_cell(task_name, TASK_NAME_WIDTH),
            config_task_run_mode(config, task),
            "-",
            "stopped",
            "-",
            0,
            "-",
            "-",
            "-",
            "-",
            "-"
        )?;
    }
    writeln!(
        output,
        "{}",
        colored_note_line(&format!("Timezone: {}", config.project.display_timezone))
    )
}

pub fn format_prefixed_logs(task: &str, logs: &str, options: &RenderLogOptions) -> String {
    let mut output = String::new();
    for line in selected_log_lines(logs, options) {
        let _ = write!(output, "{task} | {line}");
    }
    if !logs.is_empty() && !logs.ends_with('\n') {
        output.push('\n');
    }
    output
}

pub fn format_merged_logs(logs: &[(String, String)], options: &RenderLogOptions) -> String {
    let mut lines = Vec::new();
    let mut sequence = 0_usize;
    for (task, log) in logs {
        for line in selected_log_lines(log, options) {
            lines.push(LogLine {
                task: task.clone(),
                timestamp: line_timestamp(&line),
                line,
                sequence,
            });
            sequence += 1;
        }
    }
    lines.sort_by(|left, right| match (left.timestamp, right.timestamp) {
        (Some(left_ts), Some(right_ts)) => left_ts
            .cmp(&right_ts)
            .then_with(|| left.sequence.cmp(&right.sequence)),
        (Some(_), None) => std::cmp::Ordering::Less,
        (None, Some(_)) => std::cmp::Ordering::Greater,
        (None, None) => left.sequence.cmp(&right.sequence),
    });

    let mut output = String::new();
    for line in lines {
        let _ = write!(output, "{} | {}", line.task, line.line);
    }
    output
}

pub fn selected_log_lines(logs: &str, options: &RenderLogOptions) -> Vec<String> {
    let mut selected = logs
        .split_inclusive('\n')
        .filter(|line| options.grep.as_ref().is_none_or(|grep| line.contains(grep)))
        .filter(|line| {
            options.since.is_none_or(|since| {
                line_timestamp(line).is_some_and(|timestamp| timestamp >= since)
            })
        })
        .map(ToString::to_string)
        .collect::<Vec<_>>();
    if let Some(lines) = options.lines {
        selected = selected
            .into_iter()
            .rev()
            .take(lines)
            .collect::<Vec<_>>()
            .into_iter()
            .rev()
            .collect();
    }
    selected
}

pub fn line_timestamp(line: &str) -> Option<DateTime<Utc>> {
    line.split_whitespace().find_map(|token| {
        let token = token.trim_matches(|ch: char| {
            !(ch.is_ascii_alphanumeric() || matches!(ch, '-' | ':' | '.' | '+' | 'Z' | 'T'))
        });
        DateTime::parse_from_rfc3339(token)
            .ok()
            .map(|time| time.with_timezone(&Utc))
    })
}

pub fn status_label(status: TaskStatus) -> &'static str {
    match status {
        TaskStatus::Defined => "defined",
        TaskStatus::Scheduled => "scheduled",
        TaskStatus::WaitingDependency => "waiting",
        TaskStatus::Starting => "starting",
        TaskStatus::Online => "online",
        TaskStatus::Healthy => "healthy",
        TaskStatus::Unhealthy => "unhealthy",
        TaskStatus::Stopping => "stopping",
        TaskStatus::Stopped => "stopped",
        TaskStatus::Failed => "failed",
        TaskStatus::Backoff => "backoff",
        TaskStatus::Disabled => "disabled",
    }
}

pub fn colored_status_label(status: TaskStatus) -> String {
    colorize_status(status, status_label(status))
}

pub fn colored_health_label(health: &str) -> String {
    colorize_health(health, health)
}

fn colored_status_cell(status: TaskStatus) -> String {
    colorize_status(status, &format!("{:<12}", status_label(status)))
}

fn colorize_status(status: TaskStatus, value: &str) -> String {
    let color = match status {
        TaskStatus::Online | TaskStatus::Healthy => "32",
        TaskStatus::Unhealthy | TaskStatus::Failed | TaskStatus::Stopped => "31",
        TaskStatus::Starting
        | TaskStatus::Scheduled
        | TaskStatus::WaitingDependency
        | TaskStatus::Stopping => "33",
        TaskStatus::Backoff => "35",
        TaskStatus::Disabled | TaskStatus::Defined => "90",
    };
    format!("\x1b[{color}m{value}\x1b[0m")
}

fn colored_health_cell(health: &str) -> String {
    colorize_health(health, &format!("{health:<8}"))
}

fn colorize_health(health: &str, value: &str) -> String {
    match health {
        "ok" => format!("\x1b[32m{value}\x1b[0m"),
        "fail" => format!("\x1b[31m{value}\x1b[0m"),
        "-" => value.to_string(),
        _ => format!("\x1b[33m{value}\x1b[0m"),
    }
}

fn colored_restarts_cell(restart_count: u32) -> String {
    let value = format!("{restart_count:<8}");
    if restart_count >= 3 {
        format!("\x1b[33m{value}\x1b[0m")
    } else {
        value
    }
}

fn colored_cpu_cell(cpu: &str) -> String {
    let value = format!("{cpu:<8}");
    if parse_cpu_percent(cpu).is_some_and(|percent| percent >= 80.0) {
        format!("\x1b[33m{value}\x1b[0m")
    } else {
        value
    }
}

fn format_cpu_percent(cpu_percent: f64) -> String {
    format!("{cpu_percent:.1}%")
}

fn colored_memory_cell(memory_bytes: Option<u64>) -> String {
    let value = memory_bytes
        .map(format_bytes)
        .unwrap_or_else(|| "-".to_string());
    let value = format!("{value:<8}");
    if memory_bytes.is_some_and(|bytes| bytes >= 512 * 1024 * 1024) {
        format!("\x1b[33m{value}\x1b[0m")
    } else {
        value
    }
}

pub fn colored_note_line(value: &str) -> String {
    format!("\x1b[90m{value}\x1b[0m")
}

fn parse_cpu_percent(cpu: &str) -> Option<f64> {
    let value = cpu.trim().strip_suffix('%').unwrap_or(cpu.trim());
    value.parse::<f64>().ok()
}

fn config_task_run_mode(config: &ProjectConfig, task: &TaskConfig) -> &'static str {
    if task
        .schedule
        .as_ref()
        .is_some_and(|schedule| schedule.start.is_some() || schedule.stop.is_some())
    {
        return "scheduled";
    }

    if !task.cron.is_empty() {
        return "cron";
    }

    let restart_policy = task.restart.unwrap_or(config.defaults.restart);
    if restart_policy != RestartPolicy::Never || task.health.is_some() || task.watch.is_some() {
        return "long";
    }

    "oneshot"
}

fn display_run_mode(task: &TaskInfo) -> &str {
    if task.run_mode.is_empty() {
        "-"
    } else {
        &task.run_mode
    }
}

fn table_display_timezone(tasks: &[TaskInfo]) -> &str {
    tasks
        .iter()
        .find_map(|task| task.display_timezone.as_deref())
        .unwrap_or("local")
}

pub fn format_duration(uptime_ms: u64) -> String {
    let seconds = uptime_ms / 1_000;
    if seconds < 60 {
        return format!("{seconds}s");
    }
    let minutes = seconds / 60;
    if minutes < 60 {
        let seconds = seconds % 60;
        return format!("{minutes}m{seconds}s");
    }
    let hours = minutes / 60;
    if hours < 24 {
        let minutes = minutes % 60;
        return format!("{hours}h{minutes}m");
    }
    let days = hours / 24;
    let hours = hours % 24;
    format!("{days}d{hours}h")
}

pub fn format_bytes(bytes: u64) -> String {
    const KB: u64 = 1024;
    const MB: u64 = 1024 * KB;
    const GB: u64 = 1024 * MB;
    if bytes >= GB {
        return format!("{}GB", bytes / GB);
    }
    if bytes >= MB {
        return format!("{}MB", bytes / MB);
    }
    if bytes >= KB {
        return format!("{}KB", bytes / KB);
    }
    format!("{bytes}B")
}

fn fixed_width_cell(value: &str, width: usize) -> String {
    let text = if value.chars().count() > width {
        let prefix = value
            .chars()
            .take(width.saturating_sub(3))
            .collect::<String>();
        format!("{prefix}...")
    } else {
        value.to_string()
    };
    format!("{text:<width$}")
}
